### PR TITLE
Bump `json-lib` from 2.4-jenkins-2 to 2.4-jenkins-3

### PR DESCRIPTION
**Pull request by basil**
stapler/stapler#262 was merged, which means the next release of Stapler will include that PR. When Dependabot proposes that release of Stapler to Jenkins core, a `RequireUpperBoundDeps` error will be triggered for `json-lib`. So we might as well get ahead of the curve by upgrading `json-lib` in Jenkins core now.

@timja was kind enough to post the diff in the Stapler PR: https://github.com/jenkinsci/json-lib/compare/json-lib-2.4-jenkins-2...json-lib-2.4-jenkins-3

In [#5629 (comment)](https://github.com/jenkinsci/jenkins/pull/5629#discussion_r673245763), @daniel-beck made the following observation:

> We need to be extremely careful updating this library given how JSON signature validation works. I struggled with unexpectedly broken JSON serialization results a lot in update-center2.

This level of extreme care does not appear to have been taken in stapler/stapler#262, but it's still not too late to properly test this change. Unfortunately, I'm not sure offhand how to go about such testing. I need to think about this some more, so I am leaving the PR in draft state while I try to determine how this should be tested. If others have concrete and actionable suggestions, I would happily take them, though I don't have any expectations here.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]
🤷 I merged the other one because the only change I could see was string `"null"` is not the same as `null` (and `undefined`.

I can't really see that being an issue =/ but if @daniel-beck has any suggestions / checks he'd like to see that would be good

**Comment by daniel-beck** [comment]
> Unfortunately, I'm not sure offhand how to go about such testing.

The problem in general is the canonicalization applied to JSON received from update sites.

The update center builds a giant JSON object, serializes it computing its string checksum, signs that checksum, and inserts both checksum and signature (plus some extra stuff) as a `signatures` object into the JSON. The result is then serialized again and served to Jenkins.

Jenkins deserializes the giant JSON object, grabs the `signatures` object, removes it from the rest, serializes the rest computing the string checksum, and confirms it's the same as stored in `signatures`. Then it checks the signature, but that doesn't really matter here.

Basically, update-center2 and Jenkins need to serialize JSON to identical strings: Key ordering, whitespace, number formatting, etc.

IIRC, currently number formatting is inconsistent between the two components after I optimized some stuff in infra, which is why popularity is an int instead of a float/double as I had originally planned.

It's possible that this update is entirely harmless, but if we don't notice a discrepancy that'll mess with us later, we'll need to do rework on infra (serving different JSON based on Jenkins version, and not just content-wise, but different serialized formats), or accept that some releases of Jenkins don't have any update site access anymore.

**Comment by basil** [comment]
@daniel-beck Do you have a concrete and actionable suggestion about how I should test this?

**Review by oleg-nenashev — APPROVED** [review]
It definitely makes sense to minimize usage of the library and switch to stock implementations where possible. The update center part is indeed complicated and risky, but right now there are no update blockers.

I see no blockers for this PR

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jglick** [comment]
Apparently causes regressions in some plugins calling `ListBoxModel.add("display text", null)` and later expecting `@DataBoundConstructor` / `@DataBoundSetter` to receive `null` rather than `""`. This was incorrect to begin with (form values should never be null) but worked formerly somehow, probably by means of craziness such as

```html
<input value='"null"'>display text</input>
```

**Pull request merged**
